tracing-mock/README.md

```diff
@@ -0,0 +1,170 @@
+
+[splash]: https://raw.githubusercontent.com/tokio-rs/tracing/master/assets/splash.svg
+
+# tracing-mock
+
+Utilities for testing [`tracing`][tracing] and crates that uses it.
+
+[![Documentation (master)][docs-master-badge]][docs-master-url]
+[![MIT licensed][mit-badge]][mit-url]
+[![Build Status][actions-badge]][actions-url]
+[![Discord chat][discord-badge]][discord-url]
+
+[Documentation][docs-master-url] | [Chat][discord-url]
+
+[docs-master-badge]: https://img.shields.io/badge/docs-master-blue
+[docs-master-url]: https://tracing-rs.netlify.com/tracing_mock
+[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg
+[mit-url]: LICENSE
+[actions-badge]: https://github.com/tokio-rs/tracing/workflows/CI/badge.svg
+[actions-url]:https://github.com/tokio-rs/tracing/actions?query=workflow%3ACI
+[discord-badge]: https://img.shields.io/discord/500028886025895936?logo=discord&label=discord&logoColor=white
+[discord-url]: https://discord.gg/EeF3cQw
+
+## Overview
+
+[`tracing`] is a framework for instrumenting Rust programs to collect
+structured, event-based diagnostic information. This crate provides
+constructors for mock `tracing` objects to allow testing of `tracing`
+itself as well as crates that use it.
+
+*Compiler support: [requires `rustc` 1.49+][msrv]*
+
+[msrv]: #supported-rust-versions
+
+## Usage
+
+The `tracing-mock` crate provides a mock `subscriber` which can
+verify the spans and events sent. It is inteded for use in tests
+to ensure that your crate is tracing at the corrrect time with the
+correct data.
+
+As `tracing-mock` isn't available on [crates.io](https://crates.io/)
+yet, you must import it via git. It is important that you also override
+the source of any `tracing` crates that are transient dependencies. For
+example, the `Cargo.toml` for your test crate could contain:
+
+```toml
+[dependencies]
+lib-under-test = 1.0 # depends on `tracing`
+
+[dev-dependencies]
+tracing-mock = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x", version = "0.1" }
+tracing = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x", version = "0.1" }
+
+[patch.crates-io]
+tracing = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x" }
+```
+
+Below is a simple example of checking for an event with only a
+message.
+
+```rust
+use tracing::subscriber::with_default;
+use tracing_mock::{event, field, subscriber};
+
+fn yak_shaving() {
+    tracing::info!("preparing to shave yaks");
+}
+
+#[test]
+fn traced_event() {
+    let (subscriber, handle) = subscriber::mock()
+        .event(event::mock().with_fields(field::msg("preparing to shave yaks")))
+        .done()
+        .run_with_handle();
+
+    with_default(subscriber, || {
+        yak_shaving();
+    });
+
+    handle.assert_finished();
+}
+```
+
+The following is a more complex example taking the full yak shaving example
+from the `tracing` README and additionally instrumenting the called function
+with a span.
```

Comment:
```suggestion
Below is a slightly more complex example. `tracing-mock` asserts that, in order:
- a span is created with a single field/value pair and entered
- an event is created with the field `number_of_yaks`, a corresponding value of 3, and the message "preparing to shave yaks", and nothing else.
- an event is created with the field `all_yaks_shaved`, a corresponding value of `true`, and the message "yak shaving completed".
- the span was exited.
```

note, i think it'd be a good idea to assert that the span also _closed_.

Reply:
By closed, do you mean `.drop_span()`?